### PR TITLE
Add hotwords

### PR DESCRIPTION
**Pull request by FieldsMedal**

### CODE REVIEW

File: swig/test/test_zh.py [hunk 1]
```diff
@@ -104,6 +105,61 @@ def load_vocab(vocab_file):
...
+    batch_len_list = batch_lens.tolist()
+    batch_log_probs_seq = []
+    batch_log_probs_ids = []
+    batch_start = []
+    batch_root = TrieVector()
+    root_dict = {}
+    scorer = Scorer(0.5, 0.5, lm_path, vocab_list)
+    batchhotwords = BatchHotwords()
+    #In the first badcase, there is a big difference in scoring between the optimal path and other paths.
+    hot_words = {'极点': 5, '换一': -3.40282e+38, '首歌': -100, '换歌': 3.40282e+38}
```
Comment (yuekaizhang) [inline swig/test/test_zh.py:135]: Could you explain more why setting these words here? Why do you use positive score or negtive score? What's the bad case? Results before and after setting it?

Comment (FieldsMedal) [reply swig/test/test_zh.py:135]: In test_zh.py，batch_log_ctc_probs contains two audio test data. 
1）first audio,decoding without hotwords result is ('换一首歌', '', '换一', '一', '换', '换首歌', '一首歌', '一歌', '首歌', '歌'), i want to use hotwords chang the decoding result, let '换歌' ranking frontmost, so i set '换歌' positive weights 3.40282e+38 and reduce the weight of other words like "换一" and "首歌".The decoding result with hot words is ('换歌', '换一首歌', '', '换一', '一', '换', '一首歌', '一歌', '首歌', '歌'）；
2）、second audio, decoding without hotwords result is ('几点了', '极点了', '几点啊', '几点啦', '几点儿', '几点吧', '急点了', '几点晚', '极点啊', '极点啦'). After setting '极点': 5，the decoding result is （'极点了',** '几点了', '极点啊', '极点啦', '极点儿', '极点吧', '极点晚', '极点呀', '几点啊', '几点啦').

**Comment by yuekaizhang** [comment]
> 

Thanks. Would you mind uploading the decoding results w w/o hotwords somewhere? 


Also, https://www.modelscope.cn/models/damo/speech_paraformer-large-contextual_asr_nat-zh-cn-16k-common-vocab8404/summary here, they use F1 score, recall, precision to evaluate hot words. Can we get this stats also?


I am also interested about the general test set performance. Would you mind also test the normal aishell test set WER?
https://github.com/yuekaizhang/Triton-ASR-Client/blob/main/client.py#L37-L43

**Comment by FieldsMedal** [comment]
> > 
> 
> Thanks. Would you mind uploading the decoding results w w/o hotwords somewhere? (Maybe a huggingface repo for hotwords weight, ngram file, decoding results and other essentials is a good choice.)

hotwords weight and ngram file:https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/tree/main/models.
decoding results: https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/tree/main/results.

- **The current order of ngram is 4, only support length <= 4 hotwords, if you want to configure longer hotwords, you can use higher order ngram, but at the same time will increase the decoding time.**      

<br>

> Also, https://www.modelscope.cn/models/damo/speech_paraformer-large-contextual_asr_nat-zh-cn-16k-common-vocab8404/summary here, they use F1 score, recall, precision to evaluate hot words. Can we get this stats?

Hotwords result on [speech_asr_aishell1_hotwords_testsets](https://www.modelscope.cn/datasets/speech_asr/speech_asr_aishell1_hotwords_testsets/summary).
| model (FP16)               | Latency (s) | CER   | Recall | Precision | F1-score |
|----------------------------|-------------|-------|--------|-----------|----------|
| offline model w/o hotwords | 5.5921      | 13.85 | 0.27   | 0.99      | 0.43     |
| offline model w/  hotwords | 5.6401      | 12.16 | 0.45   | 0.97      | 0.62     |
<br>
> I am also interested about the general test set performance. Would you mind testing the normal aishell test set WER? https://github.com/yuekaizhang/Triton-ASR-Client/blob/main/client.py#L37-L43

Hotwords result on [AISHELL-1 Test dataset](https://www.openslr.org/33/)
| model (FP16)                           | RTF        | CER     |
|------------------------------|---------|--------|
| offline model w/o hotwords   | 0.00437 | 4.6805 |
| offline model w/  hotwords   | 0.00435 | 4.5831 |
| streaming model w/o hotwords | 0.01231 | 5.2777 |
| streaming model w/  hotwords | 0.01142 | 5.1926 |


### Tested ENV
* CPU：40 Core, Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz
* GPU：NVIDIA GeForce RTX 2080 Ti

**Comment by yuekaizhang** [comment]
> > > 
> > 
> > 
> > Thanks. Would you mind uploading the decoding results w w/o hotwords somewhere? (Maybe a huggingface repo for hotwords weight, ngram file, decoding results and other essentials is a good choice.)
> 
> hotwords weight and ngram file:https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/tree/main/models. decoding results: https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/tree/main/results.
> 
> * **The current order of ngram is 4, only support length <= 4 hotwords, if you want to configure longer hotwords, you can use higher order ngram, but at the same time will increase the decoding time.**
> 
> 
> > Also, https://www.modelscope.cn/models/damo/speech_paraformer-large-contextual_asr_nat-zh-cn-16k-common-vocab8404/summary here, they use F1 score, recall, precision to evaluate hot words. Can we get this stats?
> 
> Hotwords result on [speech_asr_aishell1_hotwords_testsets](https://www.modelscope.cn/datasets/speech_asr/speech_asr_aishell1_hotwords_testsets/summary).
> 
> model (FP16)	Latency (s)	CER	Recall	Precision	F1-score
> offline model w/o hotwords	5.5921	13.85	0.27	0.99	0.43
> offline model w/ hotwords	5.6401	12.16	0.45	0.97	0.62
> > I am also interested about the general test set performance. Would you mind testing the normal aishell test set WER? https://github.com/yuekaizhang/Triton-ASR-Client/blob/main/client.py#L37-L43
> 
> Hotwords result on [AISHELL-1 Test dataset](https://www.openslr.org/33/)
> 
> model (FP16)	RTF	CER
> offline model w/o hotwords	0.00437	4.6805
> offline model w/ hotwords	0.00435	4.5831
> streaming model w/o hotwords	0.01231	5.2777
> streaming model w/ hotwords	0.01142	5.1926
> ### Tested ENV
> * CPU：40 Core, Intel(R) Xeon(R) Silver 4210 CPU @ 2.20GHz
> * GPU：NVIDIA GeForce RTX 2080 Ti

Many thanks. This results look nice. I was wondering for both w/ or w/o hotwords case if we use this as the default external LM https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/blob/main/models/init_kenlm.arpa.

**Comment by FieldsMedal** [comment]
> Many thanks. The results look nice. I was wondering for both w/ or w/o hotwords case if we use this as the default external LM https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/blob/main/models/init_kenlm.arpa.
> 
> Also, is the pretrained model from here https://github.com/wenet-e2e/wenet/tree/main/examples/aishell/s0#u2-conformer-result ? Looks like WER with WFST Decoding + attention rescoring for offline and chunk16 are 4.4 & 4.75. Pure attention rescoring without any ngram are 4.63&5.05. Not sure the results look like if you use aishell train set as arpa. I thought they use this 3-gram arpa https://huggingface.co/yuekai/aishell1_tlg_essentials/blob/main/3-gram.unpruned.arpa here.

1. init_kenlm.arpa is used to init [Score](https://github.com/Slyne/ctc_decoder/blob/master/swig/scorer.cpp#LL58C17-L58C17), because our hotwordsboosting depends on [Scorer::make_ngram](https://github.com/Slyne/ctc_decoder/blob/68f8baebcf63c7ca04cf82fe35095af28478b641/swig/scorer.cpp#L176), any language model trained by Kenlm is fine. if decode with hotwords, put unique hot words from each recording into [batchhotwords](https://github.com/Slyne/ctc_decoder/blob/9c6f664aab9092037ed6d10a09121346da929a78/swig/test/test_zh.py#L133C5-L145). if decode without hotwords, put None into batchhotwords. whether to add hotwords score controled by [this](https://github.com/Slyne/ctc_decoder/blob/9c6f664aab9092037ed6d10a09121346da929a78/swig/ctc_beam_search_decoder.cpp#LL154C19-L154C19). if user also wants to add ngram score, set [use_ngram_score](https://github.com/Slyne/ctc_decoder/blob/9c6f664aab9092037ed6d10a09121346da929a78/swig/ctc_beam_search_decoder.cpp#L174) to true.
2. Our pretrained model from here https://github.com/wenet-e2e/wenet/blob/main/docs/pretrained_models.md, trained by aishell datasets. Our results on [speech_asr_aishell1_hotwords_testsets](https://www.modelscope.cn/datasets/speech_asr/speech_asr_aishell1_hotwords_testsets/summary) were tested using the fp16 onnx model,ctc_weight is 0.3, reverse_weight is 0.3. These settings may have some impact.

File: swig/ctc_beam_search_decoder.cpp [hunk 2]
```diff
@@ -130,9 +148,39 @@ std::vector<std::pair<double, std::vector<int>>> ctc_beam_search_decoder(
...
-            score = ext_scorer->get_log_cond_prob(ngram) * ext_scorer->alpha;
+
+            // hot words boosting
+            float hot_boost = 0.0;
+            if (hotwords_scorer != nullptr && !hotwords_scorer->hotwords_dict.empty()){
+                    std::unordered_map<std::string, float>::const_iterator iter;
+                    for (size_t index=0; index < ngram.size(); index ++ ) {
+                        std::string word = "";
+                        // character-based language model, combining chinese characters into words
+                        if (ext_scorer->is_character_based()) {
```
Comment (yuekaizhang) [inline swig/ctc_beam_search_decoder.cpp:159]: I am not fully understand why we need a external arpa file to launch the hotword scorer when we set use_ngram_score=False. If the ext_scorer->is_character_based() is the reason, how about setting an independent flag?

Comment (yuekaizhang) [reply swig/ctc_beam_search_decoder.cpp:159]: I see. So, if we use a 4-gram-arpa as ext_scorer, we can't handle hot words more than 4 like this one https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords/blob/main/models/hotwords.yaml#L4, right? 

What do you think if we copy that make_ngram into HotWord Class? Such that we could set the hot word ngram max_length as well as is_character_or_not.

Since it would be a little bit confusing, when we set the lm_path, but our final decoding results didn't use ngram lm.

Comment (FieldsMedal) [reply swig/ctc_beam_search_decoder.cpp:159]: > 

1. If 4-gram-arpa as ext_scorer，only support a maximum length of 4 Chinese characters.
2. I plan to copy make_ngram into HotWordsBoosting Class, and exposing the length parameter of hotwords.

Comment (FieldsMedal) [reply swig/ctc_beam_search_decoder.cpp]: In the latest commit, we separated hotwords scoring from language model scoring. The latest hotwords results on speech_asr_aishell1_hotwords_testsets and AISHELL-1 Test dataset can be found [here](https://huggingface.co/58AILab/wenet_u2pp_aishell1_with_hotwords).

**Comment by wabac2023** [comment]
In the latest commit, we modify batch_hotwords_scorer to hotwords_scorer. If you have free time, please help to review this pr.

**Comment by FieldsMedal** [comment]
In the latest commit, we modify batch_hotwords_scorer to hotwords_scorer. If you have free time, please help to review this pr.

**Comment by Slyne** [comment]
@FieldsMedal  Thanks!
One more question:
The output (Test hotwords boosting with word-level language models during ctc prefix beam search) for ```test_zh.py```  is
```
INFO:root:Test hotwords boosting with word-level language models during ctc prefix beam search
INFO:root:('', '一', '换', '一首', '极点晚', '几点啦', '极点', '几点', '', '几', '晚', '极')
```

Not sure if the above result is expected?

**Comment by Slyne** [comment]
Thanks again!
Really great feature! @FieldsMedal
